tfhe/src/shortint/server_key/tests/noise_level.rs

```diff
@@ -0,0 +1,288 @@
+                   predicate: &dyn Fn(NoiseLevel) -> NoiseLevel| {
+        let mut clone = ct.clone();
+        op(sk, &mut clone);
+        assert!(clone.noise_level() == predicate(ct.noise_level()));
+    };
+
+    test_fn(&ServerKey::unchecked_neg_assign, &|ct_noise| ct_noise);
+    test_fn(
+        &|sk, ct| {
+            ServerKey::unchecked_neg_assign_with_correcting_term(sk, ct);
+        },
+        &|ct_noise| ct_noise,
+    );
+
+    let acc = sk.generate_lookup_table(|_| 0);
+
+    test_fn(
+        &|sk, ct| ServerKey::apply_lookup_table_assign(sk, ct, &acc),
+        &|_| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_ct_binary_op_noise_level_propagation(sk: &ServerKey, ct1: &Ciphertext, ct2: &Ciphertext) {
+    let test_fn = |op: &dyn Fn(&ServerKey, &Ciphertext, &Ciphertext) -> Ciphertext,
+                   predicate: &dyn Fn(NoiseLevel, NoiseLevel) -> NoiseLevel| {
+        assert!(op(sk, ct1, ct2).noise_level() == predicate(ct1.noise_level(), ct2.noise_level()));
+    };
+
+    test_fn(&ServerKey::unchecked_add, &|ct1_noise, ct2_noise| {
+        ct1_noise + ct2_noise
+    });
+    test_fn(&ServerKey::unchecked_sub, &|ct1_noise, ct2_noise| {
+        ct1_noise + ct2_noise
+    });
+    test_fn(
+        &|sk, ct1, ct2| ServerKey::unchecked_sub_with_correcting_term(sk, ct1, ct2).0,
+        &|ct1_noise, ct2_noise| ct1_noise + ct2_noise,
+    );
+
+    let any_trivially_encrypted = ct1.degree.0 == 0 || ct2.degree.0 == 0;
+    test_fn(&ServerKey::unchecked_mul_lsb, &|_, _| {
+        if any_trivially_encrypted {
+            NoiseLevel::ZERO
+        } else {
+            NoiseLevel::NOMINAL
+        }
+    });
+    test_fn(&ServerKey::unchecked_mul_msb, &|_, _| {
+        if any_trivially_encrypted {
+            NoiseLevel::ZERO
+        } else {
+            NoiseLevel::NOMINAL
+        }
+    });
+
+    test_fn(&ServerKey::unchecked_div, &|_, _| NoiseLevel::NOMINAL);
+    test_fn(&ServerKey::unchecked_bitand, &|_, _| NoiseLevel::NOMINAL);
+    test_fn(&ServerKey::unchecked_bitor, &|_, _| NoiseLevel::NOMINAL);
+    test_fn(&ServerKey::unchecked_bitxor, &|_, _| NoiseLevel::NOMINAL);
+    test_fn(&ServerKey::unchecked_equal, &|_, _| NoiseLevel::NOMINAL);
+    test_fn(&ServerKey::unchecked_mul_lsb_small_carry, &|_, _| {
+        NoiseLevel::NOMINAL * 2
+    });
+    test_fn(&ServerKey::unchecked_greater, &|_, _| NoiseLevel::NOMINAL);
+    test_fn(&ServerKey::unchecked_greater_or_equal, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_less, &|_, _| NoiseLevel::NOMINAL);
+    test_fn(&ServerKey::unchecked_less_or_equal, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_not_equal, &|_, _| NoiseLevel::NOMINAL);
+    test_fn(
+        &|sk, ct1, ct2| ServerKey::unchecked_evaluate_bivariate_function(sk, ct1, ct2, |_, _| 0),
+        &|_, _| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_ct_binary_op_assign_noise_level_propagation(
```

Comment:
we don't have an equivalent for smarts ? smart would need two things, clean ciphertexts and then ciphertexts with filled degree which can be simulated by setting the ct degree to Degree(msg_mod * carry_mod - 1)

Reply:
See https://github.com/zama-ai/tfhe-rs/pull/651#discussion_r1377656150

Reply:
let's have a follow up issue if we don't have one :)